Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,270 @@
+//
+//  VersionFile.swift
+//  Carthage
+//
+//  Created by Jason Boyle on 8/11/16.
+//  Copyright © 2016 Carthage. All rights reserved.
+//
+
+import Foundation
+import Runes
+import Argo
+import Curry
+import ReactiveSwift
+import ReactiveTask
+import Result
+
+private struct CachedFramework {
+	let name: String
+	let md5: String
+	
+	static let nameKey = "name"
+	static let md5Key = "md5"
+	
+	func toJSONObject() -> Any {
+		return [
+			CachedFramework.nameKey: name,
+			CachedFramework.md5Key: md5
+		]
+	}
+}
+
+extension CachedFramework: Decodable {
+	static func decode(_ j: JSON) -> Decoded<CachedFramework> {
+		return curry(self.init)
+			<^> j <| CachedFramework.nameKey
+			<*> j <| CachedFramework.md5Key
+	}
+}
+
+private struct VersionFile {
+	let commitish: String
+	let xcodeVersion: String
+	
+	let macOS: [CachedFramework]?
+	let iOS: [CachedFramework]?
+	let watchOS: [CachedFramework]?
+	let tvOS: [CachedFramework]?
+	
+	static let commitishKey = "commitish"
+	static let xcodeVersionKey = "xcodeVersion"
+	
+	func cachesForPlatform(_ platform: Platform) -> [CachedFramework]? {
+		switch platform {
+		case .macOS:
+			return macOS
+		case .iOS:
+			return iOS
+		case .watchOS:
+			return watchOS
+		case .tvOS:
+			return tvOS
+		}
+	}
+	
+	func cachedPlatforms() -> Set<Platform> {
+		return Set(Platform.supportedPlatforms.filter { self.cachesForPlatform($0) != nil })
+	}
+	
+	func toJSONObject() -> Any {
+		var dict: [String: Any] = [
+			VersionFile.commitishKey : commitish,
+			VersionFile.xcodeVersionKey : xcodeVersion
+		]
+		for platform in Platform.supportedPlatforms {
+			if let caches = cachesForPlatform(platform) {
+				dict[platform.rawValue] = caches.map { $0.toJSONObject() }
+			}
+		}
+		return dict
+	}
+	
+	init(commitish: String, xcodeVersion: String, macOS: [CachedFramework]?, iOS: [CachedFramework]?, watchOS: [CachedFramework]?, tvOS: [CachedFramework]?) {
+		self.commitish = commitish
+		self.xcodeVersion = xcodeVersion
+		
+		self.macOS = macOS
+		self.iOS = iOS
+		self.watchOS = watchOS
+		self.tvOS = tvOS
+	}
+	
+	init?(url: URL) {
+		guard FileManager.default.fileExists(atPath: url.path),
+			let jsonData = try? Data(contentsOf: url),
+			let json = try? JSONSerialization.jsonObject(with: jsonData, options: .allowFragments),
+			let versionFile: VersionFile = Argo.decode(json) else {
+			return nil
+		}
+		self = versionFile
+	}
+	
+	func check(_ platform: Platform, commitish: String, xcodeVersion: String, rootDirectoryURL: URL) -> SignalProducer<Bool, CarthageError> {
+		guard commitish == self.commitish && xcodeVersion == self.xcodeVersion else {
+			return .init(value: false)
+		}
+		
+		let rootBinariesURL = rootDirectoryURL.appendingPathComponent(CarthageBinariesFolderPath, isDirectory: true).resolvingSymlinksInPath()
```

Comment:
The URL of the binaries folder should be passed in directly—instead of putting the logic here.